### PR TITLE
8344381: [s390x] Test failures with error: Register type is not known

### PR DESCRIPTION
**Pull request by offamitkumar**
Adds SaveLiveRegister portion for vector registers also. 

Depends on https://github.com/openjdk/jdk/pull/22190. Once that PR gets integrated, will rebase and mark it ready for review.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344381](https://bugs.openjdk.org/browse/JDK-8344381): [s390x] Test failures with error: Register type is not known (**Bug** - P3)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22197/head:pull/22197` \
`$ git checkout pull/22197`

Update a local copy of the PR: \
`$ git checkout pull/22197` \
`$ git pull https://git.openjdk.org/jdk.git pull/22197/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22197`

View PR using the GUI difftool: \
`$ git pr show -t 22197`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22197.diff">https://git.openjdk.org/jdk/pull/22197.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22197#issuecomment-2482663585)
</details>

### CODE REVIEW

**Comment by offamitkumar** [comment]
I got only test failure : `java/util/Spliterator/SpliteratorTraversingAndSplittingTest.java` which seems to generic issue [JDK-8344253](https://bugs.openjdk.org/browse/JDK-8344253)

**Comment by offamitkumar** [comment]
@RealLucy can you take a look at this one. I guess this also can be treated as trivial. It will good if you can see if the register preserved for float & general register are also okay.

File: src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp [hunk 1]
```diff
@@ -277,8 +277,19 @@ int SaveLiveRegisters::iterate_over_register_mask(IterationAction action, int of
           assert(action == ACTION_COUNT_ONLY, "Sanity");
         }
       }
-    } else if (false /* vm_reg->is_VectorRegister() */){
-      fatal("Vector register support is not there yet!");
+    } else if (vm_reg->is_VectorRegister()){
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp:280]: There should be a space before `{`.

File: src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp [hunk 2]
```diff
@@ -277,8 +277,19 @@ int SaveLiveRegisters::iterate_over_register_mask(IterationAction action, int of
...
+      VectorRegister vs_reg = vm_reg->as_VectorRegister();
+      if (vs_reg->encoding() >= Z_V2->encoding() && vs_reg->encoding() <= Z_V31->encoding()) {
+        reg_save_index += 2;
+        Register spill_addr = Z_R0;
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp:284]: Unused.

**Review by TheRealMDoerr — COMMENTED** [review]
Looks basically good. I think that the instructions can encode the maximum frame size.

File: src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp (hunk 2 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp]: Removed.

**Comment by offamitkumar** [comment]
>I think that the instructions can encode the maximum frame size.

both of the instruction support displacement up to `12bit unsigned integer`, which should give us flexibility of around 4000byte So I guess we should be fine.

**Comment by RealLucy** [comment]
> > I think that the instructions can encode the maximum frame size.
> 
> both of the instruction support displacement up to `12bit unsigned integer`, which should give us flexibility of around 4000byte So I guess we should be fine.

Instead of speculation:
- GPRs: 16 registers, 8 bytes each -> 128 bytes
- FPRs: 16 registers, 8 bytes each -> 128 bytes
- VRs: 32 registers, 16 bytes each -> 512 bytes
- In total: 768 bytes. 
This is an easy fit into uimm12.

**Review by TheRealMDoerr — COMMENTED** [review]
Thanks for the update. Now, I only wonder about the `>= Z_V2` condition. How does this fit to https://github.com/openjdk/jdk/blob/dfddbcaab886b9baa731cd748bb7f547e1903b64/src/hotspot/cpu/s390/s390.ad#L745?

**Comment by offamitkumar** [comment]
> Thanks for the update. Now, I only wonder about the `>= Z_V2` condition. How does this fit to
> 
> https://github.com/openjdk/jdk/blob/dfddbcaab886b9baa731cd748bb7f547e1903b64/src/hotspot/cpu/s390/s390.ad#L745
> 
> ?

I didn't get the question actually. You mean that why didn't preserve `Z_VR0` and `Z_VR1` ? Or why we are preserving vector register `VR16` to `VR31` ?

**Comment by TheRealMDoerr** [comment]
> > Thanks for the update. Now, I only wonder about the `>= Z_V2` condition. How does this fit to
> > https://github.com/openjdk/jdk/blob/dfddbcaab886b9baa731cd748bb7f547e1903b64/src/hotspot/cpu/s390/s390.ad#L745
> > 
> > ?
> 
> I didn't get the question actually. You mean that why didn't preserve `Z_VR0` and `Z_VR1` ? Or why we are preserving vector register `VR16` to `VR31` ?

What's the reason for choosing `Z_V2`? Why are `Z_VR0` and `Z_VR1` excluded? Do `Z_VR0` to `Z_VR15` match the FP registers? If so, why are you saving and restoring `Z_V2` to `Z_V15` twice?

**Comment by offamitkumar** [comment]
> Do Z_VR0 to Z_VR15 match the FP registers?

Yes they do override. 

> If so, why are you saving and restoring Z_V2 to Z_V15 twice?

Right, so register allocator will not use them, as it is not allowed to, So no need to preserve the 2nd-half. Right. Even if we use them manually we can preserve them ourselves.

**Comment by TheRealMDoerr** [comment]
`Z_V16` to `Z_V31` should be preserved. Also see PPC64 implementation.

**Comment by offamitkumar** [comment]
I looked ppc-abi and your changes. I want to push some additional changes:
```diff
diff --git a/src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp b/src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp
index 57d38170a4d..3fff500e47c 100644
--- a/src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp
+++ b/src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp
@@ -240,6 +240,7 @@ int SaveLiveRegisters::iterate_over_register_mask(IterationAction action, int of
   int reg_save_index = 0;
   RegMaskIterator live_regs_iterator(_reg_mask);
 
+  // Going to preserve the volatile registers which can be used by Register Allocator.
   while(live_regs_iterator.has_next()) {
     const OptoReg::Name opto_reg = live_regs_iterator.next();
 
@@ -251,8 +252,10 @@ int SaveLiveRegisters::iterate_over_register_mask(IterationAction action, int of
     const VMReg vm_reg = OptoReg::as_VMReg(opto_reg);
     if (vm_reg->is_Register()) {
       Register std_reg = vm_reg->as_Register();
-
-      if (std_reg->encoding() >= Z_R2->encoding() && std_reg->encoding() <= Z_R15->encoding()) {
+      // Z_R0 and Z_R1 will not be allocated by the register allocator, see s390.ad (Integer Register Classes)
+      // Z_R6 to Z_R15 are saved registers, except Z_R14 (see Z-Abi)
+      if (std_reg->encoding() == Z_R14->encoding()
+          || (std_reg->encoding() >= Z_R2->encoding() && std_reg->encoding() <= Z_R5->encoding())) {
         reg_save_index++;
 
         if (action == ACTION_SAVE) {
@@ -265,7 +268,8 @@ int SaveLiveRegisters::iterate_over_register_mask(IterationAction action, int of
       }
     } else if (vm_reg->is_FloatRegister()) {
       FloatRegister fp_reg = vm_reg->as_FloatRegister();
-      if (fp_reg->encoding() >= Z_F0->encoding() && fp_reg->encoding() <= Z_F15->encoding()
+      // Z_R1 will not be allocated by the register allocator, see s390.ad (Float Register Classes)
+      if (fp_reg->encoding() >= Z_F0->encoding() && fp_reg->encoding() <= Z_F7->encoding()
           && fp_reg->encoding() != Z_F1->encoding()) {
         reg_save_index++;
 
@@ -279,7 +283,8 @@ int SaveLiveRegisters::iterate_over_register_mask(IterationAction action, int of
       }
     } else if (vm_reg->is_VectorRegister()) {
       VectorRegister vs_reg = vm_reg->as_VectorRegister();
-      if (vs_reg->encoding() >= Z_V2->encoding() && vs_reg->encoding() <= Z_V31->encoding()) {
+      // Z_V0 to Z_V15 will not be allocated by the register allocator, see s390.ad (reg class z_v_reg)
+      if (vs_reg->encoding() >= Z_V16->encoding() && vs_reg->encoding() <= Z_V31->encoding()) {
         reg_save_index += 2;
         if (action == ACTION_SAVE) {
           __ z_vst(vs_reg, Address(Z_SP, offset - reg_save_index * BytesPerWord));
```

z_abi


Does these seem any good ? I guess I have preserved many already-saved-registers. Maybe we can get rid of that ?

**Comment by TheRealMDoerr** [comment]
This makes sense. Only volatile registers which are allocated need to be saved and restored. You could also use `is_volatile()` functions from the `Register` class.
The registers which are not allocated should not show up here because we get them from `live_regs_iterator.next()`. So, I think there's no need to filter them out. I had basically asked because of readability reasons. Sorry if my question was a bit confusing.

**Comment by offamitkumar** [comment]
> This makes sense. Only volatile registers which are allocated need to be saved and restored. You could also use `is_volatile()` functions from the `Register` class. The registers which are not allocated should not show up here because we get them from `live_regs_iterator.next()`. So, I think there's no need to filter them out. But you can filter them out. I have no objects. I had basically asked because of readability reasons. Sorry if my question was a bit confusing.

No I guess your question was perfectly correct. My concept for "which register needs to be preserved" was a bit on the loose. But now I got idea. So thanks for that.

**Review by TheRealMDoerr — APPROVED** [review]
Looks correct to me. I would probably not exclude F1. But if the decision not to use it for allocation is fixed, ok. @RealLucy: Please take a 2nd look.

**Comment by RealLucy** [comment]
> Looks correct to me. I would probably not exclude F1. But if the decision not to use it for allocation is fixed, ok. @RealLucy: Please take a 2nd look.

Wasn't F1 used as float function return value register? If so, exclusion is very helpful.

**Review by RealLucy — APPROVED** [review]
Looks even better after the additional changes.

**Comment by offamitkumar** [comment]
> > Looks correct to me. I would probably not exclude F1. But if the decision not to use it for allocation is fixed, ok. @RealLucy: Please take a 2nd look.
> 
> Wasn't F1 used as float function return value register? If so, exclusion is very helpful.

That is F0, I have added a screenshot up for such occasions🙃

Moreover `z_dbl_reg` class comments out `F1` register and `z_rscratch1_dbl_reg` class has not a single use. So seems like register allocator will never use `Z_F1`. That's why I excluded it.

**Comment by offamitkumar** [comment]
OCA label ????

Thanks for reviews Lutz, Martin 🙂

/integrate

**Pull request closed without merging**